File: 백준/Gold/17298. 오큰수/오큰수.py
import sys
input = sys.stdin.readline

def solution():
  N = int(input())
  nums = list(map(int, input().split()))
  answer = [-1] * N
  stack = []
  
  for i in range(N):
    while stack and nums[stack[-1]] < nums[i]:
      idx = stack.pop()
      answer[idx] = nums[i]
    stack.append(i)
  
  print(' '.join(map(str, answer)))

solution()

File: 백준/Gold/17298. 오큰수/오큰수.js
function solution(input) {
  const N = parseInt(input[0], 10);
  const nums = input[1].split(" ").map(Number);
  const answer = Array(N).fill(-1);
  const stack = [];

  for (let i = 0; i < N; i++) {
    while (stack.length && nums[stack[stack.length - 1]] < nums[i]) {
      const index = stack.pop();
      answer[index] = nums[i];
    }
    stack.push(i);
  }

  console.log(answer.join(" "));
}

const input = [];
require("readline")
  .createInterface({
    input: process.stdin,
    output: process.stdout,
  })
  .on("line", (line) => {
    input.push(line);
  })
  .on("close", () => {
    solution(input);
    process.exit();
  });
